File: python/tests/family_member_test.py
from refactoring.family_member import FamilyMember


def test_descendants_of_three_generations_back_are_listed():
    # Arrange
    a = FamilyMember("a")
    b = FamilyMember("b", [a])
    c = FamilyMember("c", [b])
    d = FamilyMember("d", [c])

    # Act
    descendants = d.get_descendants(0)

    # Assert
    assert a in descendants
    assert b in descendants
    assert c in descendants


def test_descendants_of_three_generations_have_correct_generation():
    # Arrange
    a = FamilyMember("a")
    b = FamilyMember("b", [a])
    c = FamilyMember("c", [b])
    d = FamilyMember("d", [c])

    # Act
    descendants = d.get_descendants(0)

    # Assert
    assert descendants[a] == 3
    assert descendants[b] == 2
    assert descendants[c] == 1

File: csharp/Refactoring.Tests/FamilyMemberTest.cs
using Xunit;

namespace Refactoring.Tests;

public class FamilyMemberTest
{
    [Fact]
    public void TestDescendantsOfThreeGenerationsBackAreListed()
    {
        // Arrange
        FamilyMember a = new("a");
        FamilyMember b = new("b", new() { a });
        FamilyMember c = new("c", new() { b });
        FamilyMember d = new("d", new() { c });

        // Act
        var descendants = d.GetDescendants(0);

        // Assert
        Assert.True(descendants.ContainsKey(a));
        Assert.True(descendants.ContainsKey(b));
        Assert.True(descendants.ContainsKey(c));
    }

    [Fact]
    public void TestDescendantsOfThreeGenerationsHaveCorrectGeneration()
    {
        // Arrange
        FamilyMember a = new("a");
        FamilyMember b = new("b", new() { a });
        FamilyMember c = new("c", new() { b });
        FamilyMember d = new("d", new() { c });

        // Act
        var descendants = d.GetDescendants(0);

        // Assert
        Assert.Equal(3, descendants[a]);
        Assert.Equal(2, descendants[b]);
        Assert.Equal(1, descendants[c]);
    }

    [Fact]
    public void TestDescendantsOfNGenerationsHaveCorrectGeneration()
    {
        // Arrange
        FamilyMember a = new("a");
        FamilyMember b = new("b", new() { a });
        FamilyMember c = new("c", new() { b });
        FamilyMember d = new("d", new() { c });
        FamilyMember e = new("e", new() { d });
        FamilyMember f = new("f", new() { e });
        FamilyMember g = new("g", new() { f });

        // Act
        var descendants = g.GetDescendants(0);

        // Assert
        Assert.Equal(6, descendants[a]);
        Assert.Equal(5, descendants[b]);
        Assert.Equal(4, descendants[c]);
        Assert.Equal(3, descendants[d]);
        Assert.Equal(2, descendants[e]);
        Assert.Equal(1, descendants[f]);
    }
}
